### DISCUSSION TITLE
可否为文档中的实例 plugin 提供一个简单介绍

### DISCUSSION THREAD

**Discussion by tu6ge** [web-infra-dev/rspack#6662] [Q&A]
我刚开始查看这个项目，想贡献一下代码，在查看代码的时候，[plugin介绍](https://www.rspack.dev/zh/contribute/architecture/builtin-plugin) 和 [一个例子](https://github.com/web-infra-dev/rspack/blob/7cc39cc4bb6f73791a5bcb175137ffd84b105da5/crates/rspack_plugin_ignore/src/lib.rs) 这个页，但是看不懂这俩插件是做什么用的，有没有简单的文档说明一下呢

**Comment by JSerFeng** [comment]
IgnorePlugin 中 tap 了 nmf(NormalModuleFactory) 和 cmf(ContextModuleFactory) 中的 beforeResolve hook，这两个是负责模块的创建，如果返回 false 则会忽略模块的创建，IgnorePlugin 判断模块路径是否符合你传入的正则，如果符合则会忽略该模块

**Reply by tu6ge** [reply]
tap nmf cmf 等名词可否解释一下

**Reply by JSerFeng** [reply]
tap 是 [tapable](https://github.com/webpack/tapable) 这里的

nmf: normal module factory
cmf: context module factory

module factory 是 webpack/rspack 中的概念，用来产生模块的，一般的 module 都是由 normal module factory 产生的，具体还是得看源码里面模块创建过程

**Comment by tu6ge** [comment]
这个 tap 是不是 react 里面的概念（我没搞过react）

**Comment by tu6ge** [comment]
能不能给解释下这两个函数的作用呢

```rust
#[plugin_hook(NormalModuleFactoryBeforeResolve for IgnorePlugin)]
async fn nmf_before_resolve(&self, data: &mut ModuleFactoryCreateData) -> Result<Option<bool>> {
  todo!()
}

#[plugin_hook(ContextModuleFactoryBeforeResolve for IgnorePlugin)]
async fn cmf_before_resolve(&self, data: BeforeResolveResult) -> Result<BeforeResolveResult> {
   todo!()
}
```

**Comment by tu6ge** [comment]
能否搞一个 hello world 版的插件出来，然后让其他贡献者知道，插件的运作流程是怎样的，每个步骤都起了什么作用等等（最小可行的插件）

**Comment by tu6ge** [comment]
我思来想去，感觉目前这个插件的模式太过复杂了，可以参考 std 里的 Debug trait 来实现，我目前有这么一个思路，不知是否可行
```rust
trait Plugin {
    fn apply(&self, ctx: &mut Context);
}
```
用一个上下文来来做总线，如果是插件自己的配置，可以放在插件结构体内部, 在 `Context` 上可以有好多方法，在某个插件上，可以随便调用，还可以用 trait 来扩展， context可以是这样的：
```rust
impl Context {
     fn foo(&self) {
         todo!()
     }
     fn foo2(&mut self, input: Input){
          todo!()
     }
     //...
}
```

某个插件实现的时候，可以是这样的
```rust
impl Plugin for MyPlugin {
   fn apply(&self, ctx: &mut Context){
       ctx.foo();
       ctx.foo2(self.input);
   }
}
```

再有就是，在执行插件的地方，可以是这样的：
```rust
let mut ctx = Context::new(...);
let plugin1 = MyPlugin::new(option1);
let plugin2 = MySecondPlugin::new();
plugin1.apply(&mut ctx);
plugin2.apply(&mut ctx);
```
当然了，返回值应该改成 `Result`
